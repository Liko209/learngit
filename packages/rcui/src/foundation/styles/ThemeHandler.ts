import { EventEmitter2 } from 'eventemitter2';
import options from './options.json';

type Options = {
  loadPath: (theme: string) => string;
};

class ThemeHandler extends EventEmitter2 {
  private options: Options = {
    loadPath: theme => `${options.basePath}${theme}.json`,
  };

  async applyTheme(theme: string) {
    const themeOptions = await this.read(theme);
    this.emit(options.bindEvent, themeOptions);
  }

  read(theme: string) {
    const { loadPath } = this.options;
    const url = loadPath(theme);

    return this.loadUrl(url);
  }

  async loadUrl(url: string) {
    return fetch(url)
      .then((response: Response) => {
        if (response.ok) {
          try {
            return response.json();
          } catch (e) {
            return Promise.reject(`failed parsing ${url} to json`);
          }
        }
        return Promise.reject(`failed loading ${url}`);
      })
      .catch(() => Promise.reject(`failed loading ${url}`));
  }
}
export { ThemeHandler };
export default new ThemeHandler();
