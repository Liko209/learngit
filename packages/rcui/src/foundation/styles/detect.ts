import applyTheme from './ThemeHandler';
import options from './options.json';

async function detect() {
  let theme;
  try {
    theme = (await applyTheme.read(options.brandThemeFile)) as {
      [brand: string]: {
        default: string;
        themes: string[];
      };
    };
  } catch (error) {
    console.log(error);
  }
  return theme;
}

export default detect;
