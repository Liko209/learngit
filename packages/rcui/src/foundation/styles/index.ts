export { default as ThemeProvider } from './ThemeProvider';
export { default as detect } from './detect';
export { default as themeHandler } from './ThemeHandler';
export * from './theme.d';
