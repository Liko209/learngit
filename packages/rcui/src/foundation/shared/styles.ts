import { Theme } from '../styles';
import { css, keyframes } from '../styled-components';

function ellipsis() {
  return css`
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
    word-break: break-word;
  `;
}

function lineClamp(lineNumber: number, maxHeight: number) {
  return css`
    overflow: hidden;
    text-overflow: ellipsis;
    display: -webkit-box;
    -webkit-line-clamp: ${lineNumber};
    -webkit-box-orient: vertical;
    max-height: ${maxHeight};
    word-break: break-word;
  `;
}

const rippleEnter = (theme: Theme) => keyframes`
  from {
    transform: scale(0);
    opacity: 0.1;
  }
  to {
    transform: scale(1);
    opacity: ${theme.opacity[5]};
  }
`;

export { ellipsis, lineClamp, rippleEnter };
