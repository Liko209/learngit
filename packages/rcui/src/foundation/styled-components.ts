import * as styledComponents from 'styled-components';
import {
  ThemedStyledComponentsModule,
  ThemeProps as StyledThemeProps,
} from "styled-components"; // tslint:disable-line

import { Theme } from './styles';

const {
  default: styled,
  css,
  keyframes,
  createGlobalStyle,
  withTheme,
  ThemeProvider,
  ThemeConsumer,
} = styledComponents as ThemedStyledComponentsModule<Theme>;

type ThemeProps = StyledThemeProps<Theme>;

export {
  css,
  createGlobalStyle,
  keyframes,
  ThemeProvider,
  withTheme,
  ThemeConsumer,
  ThemeProps,
};

export default styled;
