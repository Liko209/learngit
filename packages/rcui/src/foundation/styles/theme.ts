import palette from './theme/palette.json';
import typography from './theme/typography.json';
import opacity from './theme/opacity.json';
import radius from './theme/radius.json';
import spacing from './theme/spacing.json';
import zIndex from './theme/zIndex.json';
import breakpoints from './theme/breakpoints.json';
import boxShadow from './theme/boxShadow.json';

import {
  Theme as MuiTheme,
  ThemeOptions,
} from '@material-ui/core/styles/createMuiTheme';
import { Transitions } from '@material-ui/core/styles/transitions';
import { Shadows } from '@material-ui/core/styles/shadows';
import { ZIndex as MuiZIndex } from '@material-ui/core/styles/zIndex';
import { Spacing as MuiSpacing } from '@material-ui/core/styles/spacing';
import { Typography as MuiTypography } from '@material-ui/core/styles/createTypography';
import {
  Palette as MuiPalette,
  PaletteColor,
} from '@material-ui/core/styles/createPalette';

export type Palette = { [P in keyof typeof palette]: PaletteColor } &
  MuiPalette;
export type Typography = typeof typography & MuiTypography;
export type Opacity = typeof opacity;
export type Radius = typeof radius;
export type Spacing = typeof spacing & MuiSpacing;
export type ZIndex = typeof zIndex & MuiZIndex;
export type boxShadow = typeof boxShadow;
export type size = {
  width: number;
  height: number;
};

export { Transitions, Shadows };

export type Theme = {
  palette: Palette;
  typography: Typography;
  opacity: Opacity;
  radius: Radius;
  spacing: Spacing;
  zIndex: ZIndex;
  tooltip: {
    dark: string;
  };
  size: size;
  boxShadow: boxShadow;
  shape: {
    border1: string;
    border2: string;
    border3: string;
    border4: string;
  };
} & MuiTheme;

export type PaletteOptions = typeof palette;
export type TypographyOptions = typeof typography;
export type OpacityOptions = Opacity;
export type RadiusOptions = Radius;
export type SpacingOptions = typeof spacing;
export type ZIndexOptions = typeof zIndex;
export type BreakpointsOptions = typeof breakpoints;

export type ThemeOptions = {
  palette: PaletteOptions;
  typography: TypographyOptions;
  opacity: OpacityOptions;
  radius: OpacityOptions;
  spacing: SpacingOptions;
  zIndex: ZIndexOptions;
  breakpoints: BreakpointsOptions;
};
