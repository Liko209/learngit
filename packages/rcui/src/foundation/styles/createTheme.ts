import deepmerge from 'deepmerge';
import palette from './theme/palette.json';
import typography from './theme/typography.json';
import opacity from './theme/opacity.json';
import radius from './theme/radius.json';
import spacing from './theme/spacing.json';
import zIndex from './theme/zIndex.json';
import breakpoints from './theme/breakpoints.json';
import { createMuiTheme } from '@material-ui/core/styles';
import createMuiPalette from '@material-ui/core/styles/createPalette';
import { Theme, ThemeOptions, PaletteOptions } from './theme';

function createPalette(paletteOptions: PaletteOptions) {
  const { augmentColor } = createMuiPalette(paletteOptions);
  Object.keys(paletteOptions).forEach((key: keyof PaletteOptions) => {
    if (palette[key]) {
      augmentColor(paletteOptions[key]);
    }
  });

  return paletteOptions;
}

function createTheme(options: Partial<ThemeOptions> = {}) {
  const {
    palette: paletteInput = {},
    typography: typographyInput = {},
    opacity: opacityInput = {},
    radius: radiusInput = {},
    spacing: spacingInput = {},
    zIndex: zIndexInput = {},
    breakpoints: breakpointsInput = {},
  } = options;

  // suppress material-ui warning
  // @ts-ignore
  typographyInput.suppressWarning = true;

  const theme = deepmerge(
    {
      palette: createPalette(deepmerge(palette, paletteInput)),
      typography: deepmerge(typography, typographyInput),
      opacity: deepmerge(opacity, opacityInput),
      radius: deepmerge(radius, radiusInput),
      spacing: deepmerge(spacing, spacingInput),
      zIndex: deepmerge(zIndex, zIndexInput),
      breakpoints: deepmerge(breakpoints, breakpointsInput),
    },
    options,
  );

  return createMuiTheme(theme) as Theme;
}

export default createTheme;
