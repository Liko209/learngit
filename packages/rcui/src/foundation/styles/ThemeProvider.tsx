import React, { Component } from 'react';
import { MuiThemeProvider } from '@material-ui/core/styles';
import { ThemeProvider as StyledThemeProvider } from '../styled-components';
import createTheme from './createTheme';
import themeHandler from './ThemeHandler';
import options from './options.json';
import { Theme } from './theme.d';

type ThemeProviderProps = {
  themeName?: string;
  children: React.ReactNode;
};

type ThemeProviderPropsState = {
  theme: Theme | null;
};

class ThemeProvider extends Component<
  ThemeProviderProps,
  ThemeProviderPropsState
> {
  private _mounted: boolean;
  constructor(props: ThemeProviderProps) {
    super(props);
    this.state = {
      theme: null,
    };
    this.onThemeChanged = this.onThemeChanged.bind(this);
  }

  async componentDidMount() {
    this._mounted = true;
    const { themeName } = this.props;
    let themeOptions = {};
    if (themeName) {
      themeOptions = await themeHandler.read(themeName);
    }
    this.setState({
      theme: createTheme(themeOptions),
    });

    themeHandler.on(options.bindEvent, this.onThemeChanged);
  }

  componentWillUnmount() {
    this._mounted = false;
    themeHandler.offAny(this.onThemeChanged);
  }

  onThemeChanged(themeOptions: {}) {
    if (!this._mounted) return;
    this.setState({
      theme: createTheme(themeOptions),
    });
  }

  render() {
    const { theme } = this.state;
    const { children } = this.props;
    if (!theme) {
      return null;
    }

    return (
      <MuiThemeProvider theme={theme}>
        <StyledThemeProvider theme={theme}>
          {React.Children.only(children)}
        </StyledThemeProvider>
      </MuiThemeProvider>
    );
  }
}

export default ThemeProvider;
