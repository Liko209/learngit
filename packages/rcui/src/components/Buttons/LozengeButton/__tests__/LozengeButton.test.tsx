import * as React from 'react';
import { RuiLozengeButton } from '../';
import { ThemeProvider } from '../../../../foundation/styles';
import * as renderer from 'react-test-renderer';

describe('<Button />', () => {
  it('should render with loading', () => {
    const component = renderer
      .create(
        <ThemeProvider>
          <RuiLozengeButton loading={true}>Button</RuiLozengeButton>
        </ThemeProvider>,
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });
  it('should render without loading', () => {
    const component = renderer
      .create(
        <ThemeProvider>
          <RuiLozengeButton loading={false}>Button</RuiLozengeButton>
        </ThemeProvider>,
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });
  it('should render with arrow', () => {
    const component = renderer
      .create(
        <ThemeProvider>
          <RuiLozengeButton>Button</RuiLozengeButton>
        </ThemeProvider>,
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });
});
