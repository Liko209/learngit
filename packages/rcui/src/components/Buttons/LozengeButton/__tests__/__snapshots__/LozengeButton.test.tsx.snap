// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render with arrow 1`] = `
.c2 {
  border: 4px solid transparent;
  border-top-color: white;
  top: 2px;
  position: relative;
  margin: 4px 4px 4px 8px;
  display: inline-block;
  vertical-align: middle;
}

.c0 {
  font-size: 0.875rem;
  font-weight: 400;
  line-height: 20px;
  padding: 8px 20px;
  height: 32px;
  line-height: 0;
  border-radius: 32px;
  border: none;
  background-color: #616161;
  color: #fff;
  box-shadow: 0px 4px 5px -2px rgba(0,0,0,0.2),0px 7px 10px 1px rgba(0,0,0,0.14),0px 2px 16px 1px rgba(0,0,0,0.12);
  cursor: pointer;
  outline: none;
}

.c0:hover {
  background-color: #212121;
}

.c0.lozengeButtonLoading {
  pointer-events: none;
}

.c0.lozengeButtonLoading .c3 {
  margin-left: 4px;
}

.c0.lozengeButtonLoading .c1,
.c0.lozengeButtonWithArrow .c1,
.c0.lozengeButtonLoading .c3,
.c0.lozengeButtonWithArrow .c3,
.c0.lozengeButtonLoading span,
.c0.lozengeButtonWithArrow span {
  vertical-align: middle;
}

<button
  className="undefined lozengeButtonWithArrow c0"
>
  <span>
     
    Button
     
  </span>
  <span
    className="c1 c2"
    direction="down"
  />
</button>
`;

exports[`<Button /> should render with loading 1`] = `
.c2.c2 {
  color: white;
}

.c0 {
  font-size: 0.875rem;
  font-weight: 400;
  line-height: 20px;
  padding: 8px 20px;
  height: 32px;
  line-height: 0;
  border-radius: 32px;
  border: none;
  background-color: #616161;
  color: #fff;
  box-shadow: 0px 4px 5px -2px rgba(0,0,0,0.2),0px 7px 10px 1px rgba(0,0,0,0.14),0px 2px 16px 1px rgba(0,0,0,0.12);
  cursor: pointer;
  outline: none;
}

.c0:hover {
  background-color: #212121;
}

.c0.lozengeButtonLoading {
  pointer-events: none;
}

.c0.lozengeButtonLoading .c1 {
  margin-left: 4px;
}

.c0.lozengeButtonLoading .c3,
.c0.lozengeButtonWithArrow .c3,
.c0.lozengeButtonLoading .c1,
.c0.lozengeButtonWithArrow .c1,
.c0.lozengeButtonLoading span,
.c0.lozengeButtonWithArrow span {
  vertical-align: middle;
}

<button
  className="undefined lozengeButtonLoading lozengeButtonWithArrow c0"
>
  <span>
     
    Button
     
  </span>
  <div
    className="MuiCircularProgress-root-1 MuiCircularProgress-indeterminate-3 c1 c2"
    role="progressbar"
    style={
      Object {
        "height": 16,
        "width": 16,
      }
    }
  >
    <svg
      className="MuiCircularProgress-svg-6"
      viewBox="22 22 44 44"
    >
      <circle
        className="MuiCircularProgress-circle-7 MuiCircularProgress-circleIndeterminate-9"
        cx={44}
        cy={44}
        fill="none"
        r={20.2}
        strokeWidth={3.6}
        style={Object {}}
      />
    </svg>
  </div>
</button>
`;

exports[`<Button /> should render without loading 1`] = `
.c2 {
  border: 4px solid transparent;
  border-top-color: white;
  top: 2px;
  position: relative;
  margin: 4px 4px 4px 8px;
  display: inline-block;
  vertical-align: middle;
}

.c0 {
  font-size: 0.875rem;
  font-weight: 400;
  line-height: 20px;
  padding: 8px 20px;
  height: 32px;
  line-height: 0;
  border-radius: 32px;
  border: none;
  background-color: #616161;
  color: #fff;
  box-shadow: 0px 4px 5px -2px rgba(0,0,0,0.2),0px 7px 10px 1px rgba(0,0,0,0.14),0px 2px 16px 1px rgba(0,0,0,0.12);
  cursor: pointer;
  outline: none;
}

.c0:hover {
  background-color: #212121;
}

.c0.lozengeButtonLoading {
  pointer-events: none;
}

.c0.lozengeButtonLoading .c3 {
  margin-left: 4px;
}

.c0.lozengeButtonLoading .c1,
.c0.lozengeButtonWithArrow .c1,
.c0.lozengeButtonLoading .c3,
.c0.lozengeButtonWithArrow .c3,
.c0.lozengeButtonLoading span,
.c0.lozengeButtonWithArrow span {
  vertical-align: middle;
}

<button
  className="undefined lozengeButtonWithArrow c0"
>
  <span>
     
    Button
     
  </span>
  <span
    className="c1 c2"
    direction="down"
  />
</button>
`;
